### PR TITLE
LOOKDEVX-606 - Improve undo/redo capabilities

### PR DESCRIPTION
**Pull request by hodoulp**
Signed-off-by: Patrick Hodoul [Patrick.Hodoul@autodesk.com](mailto:Patrick.Hodoul@autodesk.com)

The pull request improves some existing code to provide an access to commands (without being executed). The aim is to enable command aggregation so the execution can be one single composite command (instead of several undoable commands).

Being in that code area, the pull request also adds extra information to the commands so, the DCC undoable queue can log meaningful information when undoing a command for example. Please refer to the `Ufe` pull request [#302](https://git.autodesk.com/media-and-entertainment/ufe/pull/302).

The pull request implements the `Ufe` pull request [#309](https://git.autodesk.com/media-and-entertainment/ufe/pull/309).

As the changes only consists to slightly refactor some existing code there are no associated unit tests.

### CODE REVIEW

**Review by JGamache-autodesk — DISMISSED** [review]

File: lib/mayaUsd/ufe/UsdContextOps.cpp [hunk 1]
```diff
@@ -278,6 +278,16 @@ class UsdUndoAddNewPrimAndSelectCommand : public Ufe::CompositeUndoableCommand
                 Ufe::GlobalSelection::get(), newSelection));
         }
     }
+
+#ifdef UFE_V4_FEATURES_AVAILABLE
+#if (UFE_PREVIEW_VERSION_NUM >= 4032)
+    std::string commandString() const override
+    {
+        return cmdsList().front()->commandString();
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdContextOps.cpp:286]: Is it possible for cmdsList() to be empty?

File: lib/mayaUsd/ufe/UsdContextOps.cpp [hunk 2]
```diff
@@ -278,6 +278,16 @@ class UsdUndoAddNewPrimAndSelectCommand : public Ufe::CompositeUndoableCommand
                 Ufe::GlobalSelection::get(), newSelection));
         }
     }
+
+#ifdef UFE_V4_FEATURES_AVAILABLE
+#if (UFE_PREVIEW_VERSION_NUM >= 4032)
```
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdContextOps.cpp:283]: Should this be 4034 to match your Ufe PR?

**Review by seando-adsk — APPROVED** [review]

File: lib/mayaUsd/ufe/UsdContextOps.cpp (hunk 2 above)
Comment (hodoulp) [reply lib/mayaUsd/ufe/UsdContextOps.cpp:283]: No the change is coming from a previous PR. Refer to https://git.autodesk.com/media-and-entertainment/ufe/pull/302 for details.

File: lib/mayaUsd/ufe/UsdContextOps.cpp (hunk 1 above)
Comment (hodoulp) [reply lib/mayaUsd/ufe/UsdContextOps.cpp:286]: In that specific case, it always contains two commands (the first is the action, and the second is the selection).

**Review by neilh-adsk — APPROVED** [review]

**Review by ppt-adsk — APPROVED** [review]
LGTM.  clang-format failing.

**Comment by hodoulp** [comment]
@based on @seando-adsk investigations, the build Preflight succeeded.

**Comment by hodoulp** [comment]
Note: The `ufe` branch is already merged.

**Review by ppt-adsk — APPROVED** [review]

**Review by seando-adsk — APPROVED** [review]

**Pull request merged**